extensions/gcp/tests/PutGCSObjectTests.cpp

```diff
@@ -0,0 +1,314 @@
+}
+
+TEST_F(PutGCSObjectTests, ServerGivesPermaError) {
+  auto return_permanent_error = [](ResumableUploadRequest const&) {
+    return google::cloud::StatusOr<std::unique_ptr<ResumableUploadSession>>(
+        PermanentError());
+  };
+
+  EXPECT_CALL(*put_gcs_object_->mock_client_, CreateResumableSession)
+      .WillOnce(return_permanent_error);
+  EXPECT_TRUE(test_controller_.plan->setProperty(put_gcs_object_, PutGCSObject::Bucket.getName(), "bucket-from-property"));
+  EXPECT_TRUE(test_controller_.plan->setProperty(put_gcs_object_, PutGCSObject::Key.getName(), "object-name-from-property"));
+  const auto& result = test_controller_.trigger("hello world");
+  EXPECT_EQ(0, result.at(PutGCSObject::Success).size());
+  ASSERT_EQ(1, result.at(PutGCSObject::Failure).size());
+  EXPECT_NE(std::nullopt, result.at(PutGCSObject::Failure)[0]->getAttribute(minifi_gcp::GCS_ERROR_DOMAIN));
+  EXPECT_NE(std::nullopt, result.at(PutGCSObject::Failure)[0]->getAttribute(minifi_gcp::GCS_ERROR_REASON));
+  EXPECT_EQ("hello world", test_controller_.plan->getContent(result.at(PutGCSObject::Failure)[0]));
+}
+
+TEST_F(PutGCSObjectTests, NonRequiredPropertiesAreMissing) {
+  EXPECT_CALL(*put_gcs_object_->mock_client_, CreateResumableSession)
+      .WillOnce([](const ResumableUploadRequest& request) {
+        EXPECT_FALSE(request.HasOption<gcs::MD5HashValue>());
+        EXPECT_FALSE(request.HasOption<gcs::Crc32cChecksumValue>());
+        EXPECT_FALSE(request.HasOption<gcs::PredefinedAcl>());
+        EXPECT_FALSE(request.HasOption<gcs::IfGenerationMatch>());
+        auto mock_upload_session = std::make_unique<gcs::testing::MockResumableUploadSession>();
+        EXPECT_CALL(*mock_upload_session, done()).WillRepeatedly(testing::Return(false));
+        EXPECT_CALL(*mock_upload_session, next_expected_byte()).WillRepeatedly(testing::Return(0));
+        EXPECT_CALL(*mock_upload_session, UploadChunk).WillRepeatedly(return_upload_in_progress());
+        EXPECT_CALL(*mock_upload_session, UploadFinalChunk).WillOnce(return_upload_done(request));
+        return google::cloud::make_status_or(std::unique_ptr<gcs::internal::ResumableUploadSession>(std::move(mock_upload_session)));
+      });
+  EXPECT_TRUE(test_controller_.plan->setProperty(put_gcs_object_, PutGCSObject::Bucket.getName(), "bucket-from-property"));
+  EXPECT_TRUE(test_controller_.plan->setProperty(put_gcs_object_, PutGCSObject::Key.getName(), "object-name-from-property"));
+  const auto& result = test_controller_.trigger("hello world");
+  EXPECT_EQ(1, result.at(PutGCSObject::Success).size());
+  EXPECT_EQ(0, result.at(PutGCSObject::Failure).size());
+}
+
+TEST_F(PutGCSObjectTests, Crc32cMD5LocationTest) {
+  EXPECT_CALL(*put_gcs_object_->mock_client_, CreateResumableSession)
+      .WillOnce([](const ResumableUploadRequest& request) {
+        EXPECT_TRUE(request.HasOption<gcs::Crc32cChecksumValue>());
+        EXPECT_TRUE(request.HasOption<gcs::MD5HashValue>());
```

Comment:
we could also check that the values are the same as what we set in the flow file attributes

Reply:
Good idea, I've changed it in https://github.com/apache/nifi-minifi-cpp/pull/1268/commits/0492f5b7b49235d69d23c34d3762176bd1f9448a#diff-5b9391803ca6a41c4f50d80c4d3c4aac2070a6be818887cdcb7356e8d7df7f15R181